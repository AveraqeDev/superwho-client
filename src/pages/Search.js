import React, { useContext, useMemo } from 'react';
import { Input, Button } from '../components/Utils';
import Table from '../components/Table';

import { ApiContext } from '../contexts/ApiContext';
import HeroApiService from '../services/hero-api-service';

import '../styles/Search.css';
import config from '../config';

export const SearchHeader = React.memo(function SearchHeader() {
  const { setResults, searched, setSearched, term, setTerm, setError } = useContext(ApiContext);

  const onSubmit = (e) => {
    e.preventDefault();
    const { search } = e.target;
    setError(null);
    setResults([]);
    setSearched(true);
    setTerm(search.value);
    HeroApiService.searchForHero(search.value)
      .then(res => {
        setResults(res);
      })
      .catch(setError);
  };

  return (
    <div className="SearchHeader">
      <h2>{searched ? `Results for "${term}"` : 'Search   '}</h2>
      <form className="SearchForm" onSubmit={onSubmit}>
        <Input 
          required
          name='search'
          id='SearchForm__term'
          placeholder='Search for a hero'
        />
        <Button className="SearchForm__button" type="submit">
          {'>'}
        </Button>
      </form>
    </div>
  )
});

const Search = () => {
  const { searched, results, error } = useContext(ApiContext);
  const columns = useMemo(() => config.COLUMNS, []);
  return (
    <div className="Search">
      <SearchHeader />
      {!error 
        ? 
          (searched 
            ? <Table columns={columns} data={results} /> 
            : (
              <div className="WelcomeText">
                <h2>Welcome to SuperWho!</h2>
                <p>To get started, type a keyword in the search box above!</p>
              </div>
            )
          ) 
        : <h2>{error.error}</h2>
      }
    </div>
  );
};

export default Search;