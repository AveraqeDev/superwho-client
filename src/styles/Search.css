.Search {
  margin-left: 15%;
}

.SearchHeader {
  background: #FFFFFF;
  display: flex;
  justify-content: space-between;
  width: 100%;
  padding: 0 2rem;
}

.SearchHeader > h2 {
  font-size: 30px;
  font-weight: bold;
}

.SearchForm {
  display: flex;
  justify-content: center;
  align-items: center;
}

#SearchForm__term {
  padding: 7px;
  margin-bottom: 0;
  width: 250px;
}

#SearchForm__term::placeholder {
  font-weight: bold;
}

.SearchForm__button {
  font-size: 20px;
  padding: 2px 6px;
  margin-left: .5rem;
}

.SearchForm__button:hover {
  opacity: 0.8;
}

@media screen and (max-width: 900px) {
  .Search {
    margin-left: 20%;
  }
}

@media screen and (max-width: 760px) {
  .Search {
    margin-left: 0;
  }
}

@media screen and (max-width: 480px) {
  .SearchHeader {
    flex-direction: column-reverse;
  }

  .SearchHeader > h2 {
    text-align: center;
    margin-top: 0;
  }

  .SearchForm {
    padding: 1rem;
  }

  #SearchForm__term {
    margin: 0;
  }
}