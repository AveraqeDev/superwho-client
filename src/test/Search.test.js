import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Search, { SearchHeader } from '../pages/Search';

describe('Search', () => {
  describe('<SearchHeader />', () => {
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(<SearchHeader />, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <SearchHeader />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('<Search />', () => {
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(<Search />, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <Search />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });
});