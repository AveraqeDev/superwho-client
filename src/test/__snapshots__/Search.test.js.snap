// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search <Search /> renders the UI as expected 1`] = `
<div
  className="Search"
>
  <div
    className="SearchHeader"
  >
    <h2>
      Search   
    </h2>
    <form
      className="SearchForm"
      onSubmit={[Function]}
    >
      <input
        className="Input "
        id="SearchForm__term"
        name="search"
        placeholder="Search for a hero"
        required={true}
      />
      <button
        className="Button SearchForm__button"
        type="submit"
      >
        &gt;
      </button>
    </form>
  </div>
  <h2 />
</div>
`;

exports[`Search <SearchHeader /> renders the UI as expected 1`] = `
<div
  className="SearchHeader"
>
  <h2>
    Search   
  </h2>
  <form
    className="SearchForm"
    onSubmit={[Function]}
  >
    <input
      className="Input "
      id="SearchForm__term"
      name="search"
      placeholder="Search for a hero"
      required={true}
    />
    <button
      className="Button SearchForm__button"
      type="submit"
    >
      &gt;
    </button>
  </form>
</div>
`;
